lib/marin/src/marin/processing/classification/dedupe.py

```diff
@@ -409,51 +499,34 @@ def compute_paragraph_hashes(batch: pa.RecordBatch) -> pa.RecordBatch:
         ]
         return dupekit.transform(batch, pipeline)
 
-    def _count_reduce(key: str, items: Iterator[pa.StructScalar]) -> dict[str, str] | None:
-        try:
-            head = next(items)
-            _ = next(items)
-        except StopIteration:
-            return None
-        # A dupe exists
-        return {
-            "hash": key,
-            "canonical": head["doc_id"],
-        }
-
     # first compute the full set of duplicate keys.
     duplicate_key_shards = list(
-        flow_backend(max_parallelism=config.processes).execute(
-            Dataset.from_list(input_files)
-            .flat_map(_load_batches)
-            .reshard(num_shards=config.processes)
+        backend.execute(
+            Dataset.from_list(input_files).flat_map(_load_batches)
+            # NOTE: when do we want to trigger reshard. Keep in mind that reshard will materialize the
+            #   text field!
+            # TODO: the resharding logic should be improved, based on size and/or max_parallelism
+            .reshard(num_shards=config.processes if len(input_files) > 3 and len(input_files) < 42 else None)
```

Comment:
Yeah we also probably want to invest in adding good repartition support in Zephyr so we don't have to materialize the full dataset in the Ray object store.

e.g. if we had "repartition(key_fn)" we could reshard everything to xxhash(docid) and use that consistently downstream.

Ray unfortunately makes doing the repartition in a streaming way hard. The best thing I can think to do is that you would do something like:

stage1:
  for each shard:
     stream over it, read ~1GB, reshard in memory, write to parquet in sorted order by target_shard_id

stage2:
  read all input shards, scan forward to just pulling out the shard id for this new shard.

Then you would at least only make a single copy on GCS and not e.g. input_shards * output_shards new files.